### PR TITLE
enhancement(datadog_agent source): comply with component spec

### LINKED ISSUE #10059: Instrument the `datadog_agent` source with event processing metrics
Per #9326

### PR DESCRIPTION
**Pull request by jdrouet**
Add `EventsSent` and `StreamError` metrics

Closes #10059

Signed-off-by: Jérémie Drouet <jeremie.drouet@gmail.com>

### CODE REVIEW

File: src/internal_events/datadog_agent.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+use crate::source_sender::ClosedError;
+use metrics::counter;
+use vector_core::internal_event::InternalEvent;
+
+#[derive(Debug)]
+pub struct DatadogAgentStreamError {
+    pub error: ClosedError,
+    pub count: usize,
+}
```
Comment (bruceg) [inline src/internal_events/datadog_agent.rs:9]: As indicated in my previous review, I think this should be a generic error for all sources.

File: src/internal_events/datadog_agent.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+impl InternalEvent for DatadogAgentStreamError {
+    fn emit_logs(&self) {
+        error!(
+            message = "Failed to forward events, downstream is closed.",
+            error = %self.error,
+            error_type = "stream",
```
Comment (bruceg) [inline src/internal_events/datadog_agent.rs:16]: I recommend something like `stream_closed` for these, although that may be too confusing, as it sounds like maybe _upstream_ was closed, which is not what this is describing.

File: src/sources/datadog/agent/mod.rs [hunk 3]
```diff
@@ -241,12 +244,10 @@ impl DatadogAgentSource {
...
                 }
                 .map_err(move |error: crate::source_sender::ClosedError| {
-                    // can only fail if receiving end disconnected, so we are shutting down,
-                    // probably not gracefully.
-                    error!(message = "Failed to forward events, downstream is closed.");
-                    error!(message = "Tried to send the following event.", %error);
+                    emit!(&DatadogAgentStreamError { error, count });
                     warp::reject::custom(ApiError::ServerShutdown)
                 })?;
+                emit!(&EventsSent { count, byte_size });
```
Comment (bruceg) [inline src/sources/datadog/agent/mod.rs:250]: Noting as per the other review, this is not needed for sources.

**Review by bruceg — CHANGES_REQUESTED** [review]

**Comment by jdrouet** [comment]
@bruceg all your suggestions have been addressed, could you take another look?

**Review by bruceg — APPROVED** [review]

**Review by jszwedko — COMMENTED** [review]
Again, here, I believe that the reported `component_received_event_bytes_total` is inaccurate. It appears to be reporting the raw bytes rather than the in-memory representation via `BytesSizeOf::size_of`.

I realize that you didn't change that, but part of this survey should include making sure the existing metrics are accurate 😄

**Comment by jdrouet** [comment]
@jszwedko that's quite weird, it's using `BytesSizeOf` [here](https://github.com/vectordotdev/vector/blob/master/src/sources/datadog/agent/mod.rs#L428), [here](https://github.com/vectordotdev/vector/blob/master/src/sources/datadog/agent/mod.rs#L463) and [here](https://github.com/vectordotdev/vector/blob/master/src/sources/datadog/agent/mod.rs#L534). 🤔 Do I miss something?

**Comment by jszwedko** [comment]
> @jszwedko that's quite weird, it's using `BytesSizeOf` [here](https://github.com/vectordotdev/vector/blob/master/src/sources/datadog/agent/mod.rs#L428), [here](https://github.com/vectordotdev/vector/blob/master/src/sources/datadog/agent/mod.rs#L463) and [here](https://github.com/vectordotdev/vector/blob/master/src/sources/datadog/agent/mod.rs#L534). 🤔 Do I miss something?

Ah, hmm, I thought I left this comment on the `socket` source. You are right, this one seems to be ok. Apologies for the confusion!

**Pull request merged**